core/string/ustring.cpp

```diff
@@ -5275,47 +5275,49 @@ String String::get_invalid_node_name_characters(bool p_allow_internal) {
 	return r;
 }
 
-String String::validate_node_name() const {
-	// This is a critical validation in node addition, so it must be optimized.
-	const char32_t *cn = ptr();
-	if (cn == nullptr) {
-		return String();
+int32_t String::invalid_node_name_index() const {
```

Comment:
I see you've split this into two functions to speed up the call.
I appreciate the thought, but I don't think this change is needed. 

Returning `*this` as in `validate_node_name` is near instantaneous (in the context of this function, and `set_name` in general. Try your PR without this change, I'd be surprised if it made any measurable speed difference :)

Reply:
@Ivorforce No, it's not a simple CoW thing. Constructing a new StringName is not a cheap operation. I tested not using this function and the performance was worse than master. This function is necessary.

Reply:
No, `StringName` construction is indeed not cheap! I would do something like this to avoid it:

```c++
String node_name_string = String(node_name);
String validated_node_name_string = .validate_node_name();
if (node_name_string != validated_node_name_string) {
    node_name = StringName(validated_node_name_string);
}
```

This should be just about as fast as what you currently have in your PR.

Reply:
Oh, I see, that makes sense. I did some testing and that seems to give like 70-80% the performance improvement of this PR, while having a cleaner API. Do you think it's worth going with your suggestion or with this PR as-is?

Reply:
I don't want to nitpick; if it makes such a big difference then your approach is definitely better :)

Reply:
Well, it's a small difference between your solution and mine, but a big difference between either and master.